### PR TITLE
Ensure consistency with error-handling across all handlers.

### PR DESCRIPTION
**Pull request by Mathew-Estafanous**
issue #11393 

**Problem:**
There are several instances where handlers will write directly to ``resp`` instead of returning the appropriate error and letting the handler’s wrapper write the status code and body. As discussed, this is not optimal since there isn’t consistency which also affects how each function can be tested.

**Solution:** 
As discussed, I have begun changing some of these cases to use the created custom errors instead of writing directly to ``resp``. 

I was made aware that several unit tests check the response code, however we are transitioning to returning an error so the response code or body is never written. From my understanding there will be 2 approaches to testing with this new system.
1) Check the returned error type directly and ignore the response body or code.
2) Transition the tests to go through ``ServeHTTP`` which allows for the handler’s wrapper to write the code and body to the response. This is actually directly related to #11396 which I was recently working on and currently have an open [PR](https://github.com/hashicorp/consul/pull/11499#pullrequestreview-807767260) for.

@dnephin recommended that I open a PR and discuss potential approaches before committing to a specific approach.

### CODE REVIEW

**Comment by Mathew-Estafanous** [comment]
@kisunji I’m planning on working on this next issue soon and I was wondering if you could let me know if I’m on the right track with my current approach?

Also, in my initial PR I mentioned 2 different approaches to testing these changes. (I’m not sure if my explanations made sense.) But do you have any suggestions on which approach I should take? I’m thinking of just transitioning the test cases to using ServeHTTP as well but I’m not sure if that is appropriate since I would only be changing a select few tests.

**Comment by kisunji** [comment]
Hey @Mathew-Estafanous, I'm trying to get a few more eyes and opinions from the team on this but we're a little tied up with work before the holiday season.

We'll keep you updated and try to respond within the month!

**Comment by Mathew-Estafanous** [comment]
Hey @kisunji. I saw you referencing this issue and was wondering if the team has had a chance to make a decision? If not, then that’s fine. I understand that the team might still be ramping up after the holidays.

**Comment by kisunji** [comment]
> Hey @kisunji. I saw you referencing this issue and was wondering if the team has had a chance to make a decision? If not, then that’s fine. I understand that the team might still be ramping up after the holidays.

Hey! Yeah it's still on our radar but some of the team is still away. I think it looks good to merge but will try to get one more thumbs up before doing so. Could you take a look at fixing the linting issue?

**Comment by Mathew-Estafanous** [comment]
Ya no worries! I pushed a few little changes that I thought I’d might as well add. One thing I noticed was that there is a sort of inconsistency with the testing.

It seems at times the tests expect that the response to be already written and as a result test for that. On the other hand some other tests would just test internal structs and leaves the response formatting to be tested somewhere else. 

Am I misunderstanding the intended use case of these tests?

**Comment by kisunji** [comment]
Just a heads up that there's a failing test on `main` (TestLeader_SystemMetadata_CRUD) which you can ignore

**Comment by Mathew-Estafanous** [comment]
> Just a heads up that there's a failing test on `main` (TestLeader_SystemMetadata_CRUD) which you can ignore

thanks for the heads up. I’m working on the failing test cases but I’m not sure what to do for ``TestDebugCommand_DebugDisabled``. It doesn’t seem like a test that is failing because of the change.

**Comment by kisunji** [comment]
> > Just a heads up that there's a failing test on `main` (TestLeader_SystemMetadata_CRUD) which you can ignore
> 
> Thanks for the heads up. I’m working on the failing test cases but I’m not sure what to do for `TestDebugCommand_DebugDisabled`. It doesn’t seem like a test that is failing because of the change.

Yeah if you think a test is failing for an unrelated reason, it probably is 😅 Don't worry too much about it- we'll take a look and retrigger a test if it looks to be flaky

**Comment by Mathew-Estafanous** [comment]
Ok I think the test cases and merge conflicts are fixed. Hopefully this test passes 🤞🏽 If they pass then I think the PR is ready for review.

File: agent/agent_endpoint.go
Comment (kisunji) [inline agent/agent_endpoint.go:768]: I think we're missing this error check

File: agent/agent_endpoint.go
Comment (kisunji) [inline agent/agent_endpoint.go:1237]: Is there a reason this has been moved?

Partitions are an enterprise feature and we intentionally check for this before ACLs

File: agent/http.go
Comment (kisunji) [inline agent/http.go:72]:
```suggestion
// UnauthorizedError should be returned by a handler when the request lacks valid authorization.
```

**Review by kisunji — COMMENTED** [review]
Looks good! A few minor comments

File: agent/agent_endpoint.go
Comment (Mathew-Estafanous) [reply agent/agent_endpoint.go:1237]: I honestly don’t remember moving that or why I might have done that. I’ll fix that as soon as I can.

File: agent/acl_endpoint_test.go
Comment (Mathew-Estafanous) [reply agent/acl_endpoint_test.go:124]: I think the test case uses the raw http handler and not ``ServeHTTP`` so it doesn’t end up writing anything to the response body. Should I go ahead and just transition that test to use ``ServeHTTP`` and add that block of code again?

File: agent/agent_endpoint.go
Comment (Mathew-Estafanous) [reply agent/agent_endpoint.go:821]: Yup!

File: agent/acl_endpoint_test.go
Comment (Mathew-Estafanous) [reply agent/acl_endpoint_test.go:124]: If not, I could add a test that directly checks the type of ``err`` returned by the raw handler.

Comment (kisunji) [reply agent/acl_endpoint_test.go:124]: Ah got it, I missed that. That's our different issue 😅 with ServeHTTP - I won't scope creep this PR

**Review by kisunji — APPROVED** [review]
Awesome, thanks for your continued contributions 👍 Will merge after tests finish

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/540683.

**Pull request merged**
